src/application-manager.coffee

```diff
@@ -0,0 +1,875 @@
+					@applications.executeStepAction(serviceAction('kill', service.serviceId, service, null, skipLock: true), { force })
+					.then =>
+						@applications.executeStepAction({
+							action: 'purge'
+							appId: app.appId
+							options:
+								skipLock: true
+						}, { force })
+					.then =>
+						@applications.executeStepAction(serviceAction('start', service.serviceId, null, service, skipLock: true), { force })
+					.then ->
+						res.status(200).json(Data: 'OK', Error: '')
+			.catch (err) ->
+				res.status(503).send(err?.message or err or 'Unknown error')
+		@router.use(@proxyvisor.router)
+
+module.exports = class ApplicationManager extends EventEmitter
+	constructor: ({ @logger, @config, @db, @eventTracker }) ->
+		@docker = new Docker()
+		@images = new Images({ @docker, @logger, @db })
+		@services = new ServiceManager({ @docker, @logger, @images, @config })
+		@networks = new Networks({ @docker, @logger })
+		@volumes = new Volumes({ @docker, @logger })
+		@proxyvisor = new Proxyvisor({ @config, @logger, @db, @docker, @images, applications: this })
+		@_targetVolatilePerServiceId = {}
+		@validActions = [
+			'kill'
+			'start'
+			'stop'
+			'updateReleaseId'
+			'fetch'
+			'removeImage'
+			'updateImage'
+			'killAll'
+			'purge'
+			'restart'
+			'cleanup'
+			'createNetworkOrVolume'
+			'removeNetworkOrVolume'
+		].concat(@proxyvisor.validActions)
+		@_router = new ApplicationManagerRouter(this)
+		@router = @_router.router
+
+		@images.on('change', @reportCurrentState)
+		@services.on('change', @reportCurrentState)
+
+	serviceAction: serviceAction
+	imageForService: imageForService
+	fetchAction: fetchAction
+
+	reportCurrentState: (data) =>
+		@emit('change', data)
+
+	init: =>
+		@images.cleanupDatabase()
+		.then =>
+			@services.attachToRunning()
+		.then =>
+			@services.listenToEvents()
+
+	# Returns the status of applications and their services
+	getStatus: =>
+		Promise.join(
+			@services.getStatus()
+			@images.getStatus()
+			(services, images) ->
+				apps = {}
+				dependent = {}
+				commit = null
+				# We iterate over the current running services and add them to the current state
+				# of the app they belong to.
+				_.forEach services, (service) ->
+					appId = service.appId
+					apps[appId] ?= {}
+					apps[appId].services ?= {}
+					# We only send commit if all services have the same
+					if !commit?
+						commit = service.commit
+					else if commit != service.commit
+						commit = false
+					if !apps[appId].services[service.imageId]?
+						apps[appId].services[service.imageId] = _.pick(service, [ 'status', 'releaseId' ])
+						apps[appId].services[service.imageId].download_progress = null
+					else
+						# There's two containers with the same imageId, so this has to be a handover
+						previousReleaseId = apps[appId].services[service.imageId].releaseId
+						apps[appId].services[service.imageId].releaseId = Math.max(parseInt(previousReleaseId), parseInt(service.releaseId)).toString()
```

Comment:
Why do we store ids as strings?

Reply:
Working to change that... we've been doing historically and it had the benefit of making it simpler when using the ids as keys to a map. Now we'll have to be more careful when comparing ids that might have been keys to a map (as they would've been "cast" to string when used as a key)